tests/test_function.py

```diff
@@ -665,6 +665,21 @@ def test_replace(self):
     self.assertEqual(ns.foo, function.zeros([2,3]))
     self.assertEqual(ns.default_geometry_name, 'y')
 
+  def test_pickle(self):
+    orig = function.Namespace()
+    domain, geom = mesh.unitsquare(2, 'square')
+    orig.x = geom
+    orig.v = domain.basis('std', 1)
+    orig.u = 'v_n ?lhs_n'
+    orig.f = 'cosh(x_0)'
+    pickled = pickle.loads(pickle.dumps(orig))
+    for attr in ('x', 'v', 'u', 'f'):
+      self.assertEqual(getattr(pickled, attr), getattr(orig, attr))
```

Comment:
Can you add a test for the `arg_shapes` attribute? For example

    self.assertEqual(pickled.arg_shapes['lhs'], orig.arg_shapes['lhs'])